### DISCUSSION TITLE
Microdot UI Testing with Selenium

### DISCUSSION THREAD

**Discussion by maxi07** [miguelgrinberg/microdot#184] [Q&A] [answered]
Has anybody experience with testing the webinterface with pytest and selenium?
This is my approach, however my website is not loading and selenium fails, as it can't find the site. Or does anybody use a different (working) framework?

```python
from selenium import webdriver
from selenium.webdriver.firefox.service import Service as FirefoxService
from webdriver_manager.firefox import GeckoDriverManager
from tests.microdot_asyncio_test_client import TestClient
import pytest


driver = webdriver.Firefox(service=FirefoxService(GeckoDriverManager().install()))
# https://pypi.org/project/webdriver-manager/


@pytest.fixture
def client(stub):
    stub.apply({
        'rp2': '[mock]',
        'neopixel': '[mock]',
        'machine': '[mock]',
        'network': '[mock]'
    })
    import settings
    client = TestClient(app=settings.app)
    return client


@pytest.mark.asyncio
async def test_mainpage(client):
    await client.get("/")
    driver.get("http://127.0.0.1")
```

fails with 

```
selenium.common.exceptions.WebDriverException: Message: Reached error page: about:neterror?e=connectionFailure&u=http%3A//127.0.0.1/&c=UTF-8&d=Firefox%20kann%20keine%20Verbindung%20zu%20dem%20Server%20unter%20127.0.0.1%20aufbauen.
```

**Comment by miguelgrinberg** [comment, marked as answer]
I think this is a misunderstanding on your part. Selenium cannot work through the test client. You have to run the real app for it to be able to connect. Try starting the app manually to begin. For the final solution you will need to start the app on a separate thread or process, and stop it when the tests end.

**Reply by maxi07** [reply]
that makes sense. Do you maybe have a starting point for me to make this run? What I would like to achieve is to make my microdot app run in a pytest environment and test button clicks or similar and not only api calls.

**Reply by miguelgrinberg** [reply]
This isn't exactly what you want because it is for Flask and not Microdot, but it may give you some ideas: https://github.com/miguelgrinberg/flasky/blob/master/tests/test_selenium.py

**Reply by maxi07** [reply]
hm what you suggest is the simplest way of getting microdot to run locally for testing ui? As its similar to Flask my first thought was to just use Flask directly, but I want to test the Microdot Browser UI directly.

**Reply by miguelgrinberg** [reply]
The file I referenced shows how to run Flask in a background thread, and then issue selenium tests. You will need to adapt that to use Microdot instead of Flask.

**Reply by maxi07** [reply]
this is my current effort but i cant get it to load the root page. From the debug output I can see that the server actually started on port 5000, but it cant find the page. Any idea?

```python
import re
import threading
import time
from selenium import webdriver


def test_1(stub):
    stub.apply({
        'rp2': '[mock]',
        'neopixel': '[mock]',
        'machine': '[mock]',
        'network': '[mock]'
    })
    client = None
    options = webdriver.ChromeOptions()
    client = webdriver.Chrome(options=options)
    import settings
    app = settings.app
    server_thread = threading.Thread(target=app.run(debug=True))
    server_thread.start()
    time.sleep(2)

    client.get('http://localhost:5000/')
    assert re.search('welcome', client.page_source)

```

**Reply by miguelgrinberg** [reply]
Take a more careful look at the example I linked. You are not creating the thread correctly. It should be:

```python
threading.Thread(target=app.run, kwargs={'debug': False})
```

**Reply by maxi07** [reply]
Unfortunately, this brings the same output. When setting `debug` to `True`, I can see that the server somehow is starting:
`Starting async server on 0.0.0.0:5000...` But then when trying to access the page, it doesn't find the webserver.
I also tried using `127.0.0.1` instead of `0.0.0.0` or different ports but with no result.

**Reply by maxi07** [reply]
So it seems like the webserver is running due to the message, but maybe it has trouble finding the routes. In my `settings.py` I am defining the Microdot server as such `app = Microdot()`. Then in my file modules/routes.py I am defining the routes such as
```python
@settings.app.get("/")
async def root_get(request):
    log_info("Loading root site.")
    try:
        log_info("Sending root site to client")
        return send_file('html/wifi-setup.html', max_age=86400)
    except Exception as ex:
        log_error("Failed sending response to client.", ex)
        return f"Failed loading main page. {ex}", 500, {'Content-Type': 'text/html'}
``` 
This works fine when running Microdot. But maybe there is an issue with the routes and my approach to start Microdot with Python?

**Reply by miguelgrinberg** [reply]
Does your test work well if you manually start the server on another terminal? Because you don't seem sure that the web server isn't running, right? Something isn't working there, but if you get the message that the server started my guess is that the issue is on the selenium side.

**Reply by maxi07** [reply]
No it doesnt seem like the server is running properly. I have tried setting up a minimal example of getting to run a very simple Microdot webserver in Python. It spins up the server and prints `Starting async server on 0.0.0.0:5000...` but its not accessible.

```python
from modules.Microdot.microdot_asyncio import Microdot
import asyncio

app = Microdot()

async def runserver():
    await app.start_server(debug=True)
    while True:
        await asyncio.sleep(0.1)

@app.get("/")
async def test(request):
    return "Hello"

asyncio.run(runserver())
```

**Reply by miguelgrinberg** [reply]
I suggest you start by trying the examples in this repository. If the above code does not work for you, I think there is something there that isn't right.

**Reply by maxi07** [reply]
When running your example code with normal python, I get the same result. The server spins up and prints the message, but it terminates instantly. But the documentation of `app.run()` normally states, that the server enters an endless listening loop, which doesn't seem to happen for me. 
https://github.com/miguelgrinberg/microdot/blob/7a329d98a8836e302109d21900288815be9fd3cf/examples/hello/hello_async.py#L33

**Reply by miguelgrinberg** [reply]
@maxi07 I can't really diagnose this for you with the little information you provided, but the `app.run()` function is not supposed to return immediately. If you have made any changes to the microdot source files I suggest you install with pip (which would work fine for CPython), to make sure you are using unmodified code.

**Comment by maxi07** [comment]
I figured it out and published my sample repo. If anyone stumbles across end to end testing with this, feel free to head over to [my sample repository](https://github.com/maxi07/microdot-playwright-pytest) on running click tests using the latest Microdot version.

@miguelgrinberg feel free to mention this in your documentation 😊
